Strap a `RequestSessionManager` to each instance of http providers

fselmo opened the pull request:
### What was wrong?

Though #3265 is not reproducible, this hopefully closes #3265 as there should be no more conflicts across separate instances due to unique instantiation of session managing per provider.

### How was it fixed?

- Use a unique session manager for each instance of HTTPProvider, and AsyncHTTPProvider, by moving all the shared session cache logic into a class and strapping it to each provider instance.
- Update tests related to session caching
- Update docs and update the typing for `request_retry_configuration` along the way (`Optional`)

### Todo:

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![20240528_215358~2](https://github.com/ethereum/web3.py/assets/3532824/ea75d536-536d-4db0-820f-24439253f4c7)

fselmo committed:
Strap a RequestSessionManager to each instance of http providers

fselmo committed:
Clean up old versions of session cache tests

fselmo committed:
Docs updates for session caching; exception retry tweaks along the way.

kclowes commented on docs/providers.rst:
```diff
@@ -107,7 +107,7 @@ local and remote JSON-RPC servers.
 ~~~~~~~~~~~~
 
-.. py:class:: web3.providers.rpc.HTTPProvider(endpoint_uri[, request_kwargs, session])
+.. py:class:: web3.providers.rpc.HTTPProvider(endpoint_uri, request_kwargs={}, session=None, exception_retry_configuration=ExceptionRetryConfiguration())
```
I believe the default of `request_kwargs` is `None`
```suggestion
.. py:class:: web3.providers.rpc.HTTPProvider(endpoint_uri, request_kwargs=None, session=None, exception_retry_configuration=ExceptionRetryConfiguration())
```

kclowes commented on docs/providers.rst:
```diff
@@ -180,7 +185,7 @@ IPCProvider
 ~~~~~~~~~~~~~~~~~
 
-.. py:class:: web3.providers.async_rpc.AsyncHTTPProvider(endpoint_uri[, request_kwargs])
+.. py:class:: web3.providers.rpc.AsyncHTTPProvider(endpoint_uri, request_kwargs={}, exception_retry_configuration=ExceptionRetryConfiguration())
```
same here:
```suggestion
.. py:class:: web3.providers.rpc.AsyncHTTPProvider(endpoint_uri, request_kwargs=None, exception_retry_configuration=ExceptionRetryConfiguration())
```

fselmo replied:
If it's none we set it to `{}` in the `__init__()` though... do we think it's better to hint at the type here by using `{}` or to be technically correct and use `None` even though that it doesn't give a good type hint?

fselmo replied:
We can use `None` and clarify below that it sets it to an empty dict? I'm ok with either approach.

kclowes replied:
Eh, I guess since it's just in the docs, we can leave it as a dict. I'll leave it up to you :)

fselmo replied:
I'll leave my reasoning for keeping it here. We have to do a dance for `ExceptionRetryConfiguration` as well where we use `Empty` since we want the ability to set it to `None` to turn it off. So I think if we go the technically correct route we'd have to put `Empty` as the default value for `ExceptionRetryConfiguration` and then explain why and I think that complicates everything more than any user cares about 😆 .

I'm going to opt to use values that if you don't touch the configuration at all, those are what we will hard code in the `__init__()`, which feels more valuable to a user than what the signature represents.

kclowes commented on web3/_utils/module_testing/module_testing_utils.py:
```diff
@@ -152,13 +146,12 @@ async def _mock_specific_request(
 
         # mock response only to specified url while validating appropriate fields
         if url_from_args == mocked_request_url:
-            assert kwargs["timeout"] == ClientTimeout(30)
```
I'm curious about this removal and the one on line 25: `assert kwargs["timeout"] == 30`. Can we still pass a timeout here?

fselmo replied:
Yeah I think that's reasonable to centralize the `DEFAULT_TIMEOUT` and use it for all http requests. I will add that back for the ccip read requests

kclowes reviewed (approved):
The idea of strapping this manager to a provider is cracking me up for some reason. I'm imagining ratchet straps 😆 Aaaanyway, this looks good to me! There wasn't really a good place to put this in the PR, but what do you think about moving the name of `web3/_utils/request.py` to something like `web3/_utils/request_session_manager.py` since the `RequestSessionManager` class is basically all the file has now? Otherwise I left some small nits, but nothing big.

reedsa commented on docs/internals.rst:
``````````````````````````````````diff
@@ -144,6 +144,29 @@ options on the provider instance:
 Retry Requests for HTTP Providers
 `````````````````````````````````
 
+.. py:class:: web3.providers.rpc.utils.ExceptionRetryConfiguration
``````````````````````````````````
In the text following this addition, it says configuration is done on the provider. I thought that it was talking about configuration for how retries are implemented and got confused.

Edit: I reread it and maybe I was confused without the additional context. Might add an intro blurb or move this below the text I mentioned.

fselmo commented:
> The idea of strapping this manager to a provider is cracking me up for some reason. I'm imagining ratchet straps 😆

😆 

> Aaaanyway, this looks good to me! There wasn't really a good place to put this in the PR, but what do you think about moving the name of `web3/_utils/request.py` to something like `web3/_utils/request_session_manager.py` since the `RequestSessionManager` class is basically all the file has now? Otherwise I left some small nits, but nothing big.

I think I'll still put the `DEFAULT_REQUEST_TIMEOUT` here and use it for the CCIP read http requests, which it doesn't matter what provider we are using, it still makes a http GET or POST as part of the spec. For that reason it might make sense to still keep this as request utils? Or should I move the class definition into `request_session_manager.py` and keep the minimal `request.py` as a space for adding more than just the `DEFAULT_REQUEST_TIMEOUT`?

fselmo committed:
docs clarification updates

kclowes commented:
> I think I'll still put the DEFAULT_REQUEST_TIMEOUT here and use it for the CCIP read http requests, which it doesn't matter what provider we are using, it still makes a http GET or POST as part of the spec. For that reason it might make sense to still keep this as request utils? Or should I move the class definition into request_session_manager.py and keep the minimal request.py as a space for adding more than just the DEFAULT_REQUEST_TIMEOUT?

Keeping DEFAULT_REQUEST_TIMEOUT in there sounds good to me, and therefore keeping this as request utils also probably makes the most sense? I wouldn't be opposed to splitting it out either though. Whatever you think!

fselmo commented:
> Keeping DEFAULT_REQUEST_TIMEOUT in there sounds good to me, and therefore keeping this as request utils also probably makes the most sense? I wouldn't be opposed to splitting it out either though. Whatever you think!

Actually, we have a *_utils/http.py* already... I'm thinking that since this is really only managing HTTP sessions, we rename the class to `HTTPSessionManager`, create a *_utils/http_session_manager.py* and put it there, and then move the default timeout to *_utils/http.py* as `HTTP_REQUEST_TIMEOUT`. Thoughts there?

kclowes commented:
Yeah, I like that. More specific!

fselmo committed:
More specific naming for session manager:

- ``RequestSessionManager`` -> ``HTTPSessionManager``
- Move default timeout for http from ``_utils/request.py`` -> ``_utils/http.py``.
  This eliminates the need for a ``request.py`` utils since ``HTTPSessionManager``
  lives in its own file at ``_utils/http_session_manager.py```.
- Pass the configured timeout to sync and async ``cache_and_return_session`` methods.
  Previously, we were using the hard-coded default value and if the call is
  coming from a provider (99% of the time), rather than the request method
  being invoked directly, we would be using the wrong value when spinning off
  any session closing tasks.

fselmo commented:
> Yeah, I like that. More specific!

Sweet. The changes were significant enough. Can I ask for one last pass starting at [this commit](https://github.com/ethereum/web3.py/pull/3412/commits/f9352e0a42c53772677d82af82d3dc93cd60f635) before we call it done? I also added passing the ``timeout`` through to the respective sync and async `cache_and_return_session()` methods so we can use a more accurate value for triggering the session closing tasks in that method.

---
edit

Just kidding I had to update the commit hash, should be good now. I added the timeout check to the integrations tests as well.

fselmo committed:
Put back timeout check on ccip read tests; close sessions appropriately.

kclowes commented on web3/_utils/http_session_manager.py:
```diff
@@ -80,7 +86,7 @@ def cache_and_return_session(
                 )
             threading.Timer(
-                DEFAULT_TIMEOUT + 0.1,
+                request_timeout + 0.1,
```
If this is None like the type allows, I think this line will throw?  Maybe add a `if request_timeout is None` check? But I'm also surprised mypy didn't catch that so maybe I missed something.

fselmo replied:
Good eye... I'll make sure to address it.

kclowes reviewed (approved):
I left one comment about a None check, but other than that, looks good!

fselmo committed:
In case ``request_timeout`` is ``None``, use ``0``, where relevant.

- Improve the async session closing by utilizing a non-blocking asyncio
  ``Task``.

fselmo committed:
Use the default http timeout to close a session if None
